src/halmos/sevm.py

```diff
@@ -2381,14 +2392,18 @@ def finalize(ex: Exec):
                         ex.st.memory.set_slice(loc, end_loc, codeslice)
 
                 elif opcode == EVM.EXTCODEHASH:
-                    account_addr = uint160(ex.st.pop())
-                    alias_addr = self.resolve_address_alias(ex, account_addr)
+                    account_addr = uint160(ex.st.peek())
+                    alias_addr = self.resolve_address_alias(
+                        ex, account_addr, stack, step_id
+                    )
+                    ex.st.pop()
+
                     addr = alias_addr if alias_addr is not None else account_addr
 
                     account_code: Optional[Contract] = ex.code.get(addr, None)
 
                     codehash = (
-                        f_extcodehash(addr)
+                        EMPTY_KECCAK
```

Comment:
is this correct?

Reply:
what? so, extcodehash of a contract account with empty code is different from that of a non-existing account?

so, here we should return 0, based on the current implementation of vms?

Reply:
> what? so, extcodehash of a contract account with empty code is different from that of a non-existing account?

yes, https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1052.md#specification is a bit more explicit

>>> The EXTCODEHASH of the account without code is c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 what is the keccack256 hash of empty data.
The EXTCODEHASH of non-existent account is 0.
The EXTCODEHASH of a precompiled contract is either c5d246... or 0.

Reply:
> so, here we should return 0, based on the current implementation of vms?

we need to have a more proper `is_empty` check for the account, which is also affected by the balance. This needs to pass:

```solidity
    function testAccountWithOrWithoutBalance() public {
        address a = address(0x42);
        assertEq(extcodehash(a), bytes32(0), "empty account has extcodehash 0");

        vm.deal(a, 1 ether);
        assertEq(extcodehash(a), 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470, "account with balance has extcodehash keccak('')");
    }
```

Reply:
full test:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import "forge-std/Test.sol";

contract Empty {
    constructor() {
        assembly {
            return(0, 0)
        }
    }
}

contract EmptyDeployerTest is Test {
    function extcodehash(address _addr) public view returns (bytes32) {
        bytes32 hash;
        assembly {
            hash := extcodehash(_addr)
        }
        return hash;
    }

    function testEmptyDeployer() public {
        address a = vm.computeCreateAddress(address(this), vm.getNonce(address(this)));
        assertEq(extcodehash(a), bytes32(0), "empty account has extcodehash 0");
        assertEq(vm.getNonce(a), 0);

        // do deploy
        new Empty();

        assertEq(a.code.length, 0);
        assertEq(extcodehash(a), keccak256(""));
        assertEq(vm.getNonce(a), 1);
    }


    function testAccountWithOrWithoutBalance() public {
        address a = address(0x42);
        assertEq(extcodehash(a), bytes32(0), "empty account has extcodehash 0");

        vm.deal(a, 1 ether);
        assertEq(extcodehash(a), keccak256(""), "account with balance has extcodehash keccak('')");
    }
}
```

in our case if we don't want to model nonces, we can probably differentiate between not being in `code`, vs being in `code` with empty bytecode